helper/logging/logfile.go

```diff
@@ -38,19 +52,89 @@ func (l *LogFile) Write(b []byte) (n int, err error) {
 			return 0, err
 		}
 	}
-
+	// Check for the last contact and rotate if necessary
+	if err := l.rotate(); err != nil {
+		return 0, err
+	}
+	l.bytesWritten += int64(len(b))
 	return l.fileInfo.Write(b)
 }
 
+func (l *LogFile) fileNamePattern() string {
+	// Extract the file extension
+	fileExt := filepath.Ext(l.fileName)
+	// If we have no file extension we append .log
+	if fileExt == "" {
+		fileExt = ".log"
+	}
+	// Remove the file extension from the filename
+	return strings.TrimSuffix(l.fileName, fileExt) + "-%s" + fileExt
+}
+
 func (l *LogFile) openNew() error {
-	newFilePath := filepath.Join(l.logPath, l.fileName)
+	fileNamePattern := l.fileNamePattern()
 
-	// Try to open an existing file or create a new one if it doesn't exist.
-	filePointer, err := os.OpenFile(newFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o640)
+	createTime := now()
+	newFileName := fmt.Sprintf(fileNamePattern, strconv.FormatInt(createTime.UnixNano(), 10))
+	newFilePath := filepath.Join(l.logPath, newFileName)
+
+	// Try creating a file. We truncate the file because we are the only authority to write the logs
+	filePointer, err := os.OpenFile(newFilePath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o640)
 	if err != nil {
 		return err
 	}
 
+	// New file, new 'bytes' tracker, new creation time :) :)
 	l.fileInfo = filePointer
+	l.lastCreated = createTime
+	l.bytesWritten = 0
+	return nil
+}
+
+func (l *LogFile) rotate() error {
+	// Get the time from the last point of contact
+	timeElapsed := time.Since(l.lastCreated)
+	// Rotate if we hit the byte file limit or the time limit
+	if (l.bytesWritten >= int64(l.maxBytes) && (l.maxBytes > 0)) || timeElapsed >= l.duration {
+		if err := l.fileInfo.Close(); err != nil {
+			return err
+		}
+		if err := l.pruneFiles(); err != nil {
+			return err
+		}
+		return l.openNew()
+	}
+	return nil
+}
+
+func (l *LogFile) pruneFiles() error {
+	if l.maxFiles == 0 {
+		return nil
+	}
+
+	pattern := filepath.Join(l.logPath, fmt.Sprintf(l.fileNamePattern(), "*"))
+	matches, err := filepath.Glob(pattern)
```

Comment:
This makes me a little uncomfortable.  Can we ensure before we get to this point that  logPath and fileNamePattern won't contain any globbing chars?

Reply:
Would `*` be sufficient, or `'?', '*', '['`? Do you know if we have any handy library func that will parse a string and return whether it contains 'globbiness'? :)